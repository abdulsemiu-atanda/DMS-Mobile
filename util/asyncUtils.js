import sendRequest from '../requests/requestUtil'

/**
 * Constructs async actions constants
 * @param {string} baseName
 * @returns {Object}
 */
export const asyncActionNames = baseName => ({
  failure: `${baseName}_FAILURE`,
  loading: `${baseName}_LOADING`,
  success: `${baseName}_SUCCESS`
})

/**
 * Constructs Redux async actions
 * @param {string} actionName
 * @returns {Object}
 */
export const asyncActions = (actionName) => ({
  loading: bool => ({
    type: asyncActionNames(actionName).loading,
    data: bool
  }),
  failure: (bool, error) => ({
    type: asyncActionNames(actionName).failure,
    data: {error, status: bool}
  }),
  success: data => ({
    type: asyncActionNames(actionName).success,
    data
  })
})

/**
 * Dispatches Redux Actions
 * @async asyncRequest
 * @param {string} ACTION_NAME - The name of action to be dispatched
 * @param {string} endpoint - The endpoint to resource
 * @param {string} method - HTTP method
 * @param {Object} data - HTTP request body
 * @param {string} token - Authentication token for request
 * @returns {Function} - Function that dispatches Redux Actions
 */

export const asyncRequest = (ACTION_NAME, endpoint, method, data, token) => dispatch => {
  dispatch(asyncActions(ACTION_NAME).loading(true))
  sendRequest(endpoint, method, data, token)
    .then(res => res.json())
    .then(res => {
      dispatch(asyncActions(ACTION_NAME).success(res))
      dispatch(asyncActions(ACTION_NAME).loading(false))
    })
    .catch(err => dispatch(asyncActions(ACTION_NAME).failure(true, err)))
}
