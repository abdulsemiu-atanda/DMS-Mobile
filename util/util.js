import {
  emailCorrect,
  emailError,
  fullnameCorrect,
  fullnameError,
  passwordMatch,
  passwordMatchError,
  usernameCorrect,
  usernameError
} from '../actions/formActions'

export const validateFields = data => {
  return dispatch => {
    Object.keys(data).forEach(field => {
      switch (field) {
        case 'email':
          return /^\S+@\S+$/g.test(data[field]) ? dispatch(emailCorrect(true)) : dispatch(emailError(true))
        case 'username':
          return data[field].trim() === '' ? dispatch(usernameError(true)) : dispatch(usernameCorrect(true))
        case 'password':
          return data.password === data.confirmPassword ? dispatch(passwordMatch(true)) : dispatch(passwordMatchError(true))
        case 'fullname':
          return data[field].trim() === '' ? dispatch(fullnameError(true)) : dispatch(fullnameCorrect(true))
        default:
          break
      }
    })
  }
}