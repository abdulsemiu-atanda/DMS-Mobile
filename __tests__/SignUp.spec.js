import {Animated} from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import {SignUp} from '../components/SignUp.react'
import {noop} from '../util/util'

const animatedValue = new Animated.Value(0)
const props = {
  asyncRequest: noop,
  frontOpacity: animatedValue.interpolate({
    inputRange: [89, 90],
    outputRange: [1, 0]
  }),
  frontInterpolate: animatedValue.interpolate({
    inputRange: [0, 180],
    outputRange: ['0deg', '180deg']
  }),
  flipCard: noop,
  goToHome: noop,
  user: {
    signingUp: false,
    message: ''
  }
}

describe('SignUp', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<SignUp {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children[1].children[0].type).toEqual('Text')
    expect(tree.children[1].children[0].children[0]).toEqual('Sign Up')
  })

  it('renders correct message when there is an error', () => {
    const failedSignUpProps = {
      ...props,
      user: {signingUp: false, message: 'Invalid Email'}
    }
    const tree = renderer.create(<SignUp {...failedSignUpProps} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children[1].children[12].type).toEqual('Text')
    expect(tree.children[1].children[12].children[0]).toEqual(failedSignUpProps.user.message)
  })

  it('renders spinner when sign up is loading', () => {
    const signingUpProps = {
      ...props,
      user: {signingUp: true, message: ''}
    }
    const tree = renderer.create(<SignUp {...signingUpProps} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children[1].children[11].children[0].type).toEqual('ActivityIndicator')
  })
})
