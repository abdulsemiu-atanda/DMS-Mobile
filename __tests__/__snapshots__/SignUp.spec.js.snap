// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders correct message when there is an error 1`] = `
<View
  style={
    Object {
      "backfaceVisibility": "hidden",
      "flex": 1,
      "height": 1334,
      "opacity": undefined,
      "position": "absolute",
      "transform": Array [
        Object {
          "rotateY": undefined,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 133.4,
        "opacity": 1,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#cfcfd1",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "center",
          "fontSize": 18,
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      Sign Up
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      FULL NAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="Jason Bourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      USERNAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="jbourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      EMAIL
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      keyboardType="email-address"
      placeholder="jason@bourne.com"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      CONFIRM PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onEndEditing={[Function]}
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0983ff",
          "height": 40,
          "marginTop": 20,
          "padding": 7,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#aacfff",
            "fontSize": 18,
          }
        }
      >
        Create
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "marginTop": 10,
        }
      }
    >
      Invalid Email
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 30,
        },
        Object {
          "height": "16%",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 12,
          "fontWeight": "600",
          "marginRight": 5,
        }
      }
    >
      ALREADY HAVE AN ACCOUNT?
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "700",
          }
        }
      >
        SIGN IN
      </Text>
    </View>
  </View>
</View>
`;

exports[`SignUp renders correctly 1`] = `
<View
  style={
    Object {
      "backfaceVisibility": "hidden",
      "flex": 1,
      "height": 1334,
      "opacity": undefined,
      "position": "absolute",
      "transform": Array [
        Object {
          "rotateY": undefined,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 133.4,
        "opacity": 1,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#cfcfd1",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "center",
          "fontSize": 18,
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      Sign Up
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      FULL NAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="Jason Bourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      USERNAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="jbourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      EMAIL
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      keyboardType="email-address"
      placeholder="jason@bourne.com"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      CONFIRM PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onEndEditing={[Function]}
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0983ff",
          "height": 40,
          "marginTop": 20,
          "padding": 7,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#aacfff",
            "fontSize": 18,
          }
        }
      >
        Create
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 30,
        },
        Object {
          "height": "16%",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 12,
          "fontWeight": "600",
          "marginRight": 5,
        }
      }
    >
      ALREADY HAVE AN ACCOUNT?
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "700",
          }
        }
      >
        SIGN IN
      </Text>
    </View>
  </View>
</View>
`;

exports[`SignUp renders spinner when sign up is loading 1`] = `
<View
  style={
    Object {
      "backfaceVisibility": "hidden",
      "flex": 1,
      "height": 1334,
      "opacity": undefined,
      "position": "absolute",
      "transform": Array [
        Object {
          "rotateY": undefined,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 133.4,
        "opacity": 1,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#cfcfd1",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "padding": 30,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "center",
          "fontSize": 18,
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      Sign Up
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      FULL NAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="Jason Bourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      USERNAME
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      placeholder="jbourne"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      EMAIL
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      keyboardType="email-address"
      placeholder="jason@bourne.com"
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 10,
        }
      }
    >
      CONFIRM PASSWORD
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onEndEditing={[Function]}
      secureTextEntry={true}
      style={
        Object {
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0983ff",
          "height": 40,
          "marginTop": 20,
          "padding": 7,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#01f0b3"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 30,
        },
        Object {
          "height": "16%",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#96a1ae",
          "fontSize": 12,
          "fontWeight": "600",
          "marginRight": 5,
        }
      }
    >
      ALREADY HAVE AN ACCOUNT?
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 12,
            "fontWeight": "700",
          }
        }
      >
        SIGN IN
      </Text>
    </View>
  </View>
</View>
`;
